src/script/interpreter.cpp

```diff
@@ -1022,6 +1025,142 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    }
+                    if (!proofStream.empty()) {
+                        // Extra bytes remaining after the MerkleProof
+                        // was deserialized, which could be a source
+                        // of witness malleability.
+                        return set_error(serror, SCRIPT_ERR_BAD_DECODE_ARG3);
+                    }
+                    if (branch.m_proof.m_path.dirty()) {
+                        // Extra bits in the final byte of the packed
+                        // serialization of the Merkle branch's path,
+                        // which would otherwise be another source of
+                        // witness malleability.
+                        return set_error(serror, SCRIPT_ERR_BAD_DECODE_ARG3);
+                    }
+                    if ((!branch.m_proof.m_path.empty() || count || !branch.m_proof.m_skip.empty()) &&
+                        ((count + branch.m_proof.m_skip.size()) != (branch.m_proof.m_path.size() + 1)))
+                    {
+                        // It is a property of any binary tree that
+                        // the number of leaf nodes is precisely one
+                        // more than the number of internal nodes.
+                        // This acts as an early-out check of whether
+                        // this is a well-formed proof. Note that the
+                        // special case of a 0-node, 0-verify, 0-skip
+                        // tree is exempted from this requirement.
+                        return set_error(serror, SCRIPT_ERR_BAD_DECODE_ARG3);
+                    }
+
+                    // The remaining _count_ items on the stack are
+                    // the verify hashes, or the actual leaf values
+                    // which are hashed with double-SHA256 to get the
+                    // verify hashes if _prehashed_ is clear.
+                    branch.m_verify.reserve(count);
+                    for (int i = 0; i < (int)count; ++i) {
+                        // -1 through -3 are the count+prehashed, root
+                        // hash, and MerkleProof we already extracted.
+                        valtype& vchLeaf = stacktop(-4 - i);
+                        if (prehashed) {
+                            // Require 32-byte hash values, no
+                            // truncation of ending bytes.
+                            if (vchLeaf.size() != 32) {
+                                return set_error(serror, SCRIPT_ERR_BAD_DECODE_ARG);
+                            }
+                            branch.m_verify.emplace_back(uint256(vchLeaf));
+                        } else {
+                            branch.m_verify.emplace_back();
+                            CHash256().Write(vchLeaf.data(), vchLeaf.size()).Finalize(branch.m_verify.back().begin());
+                        }
+                    }
+
+                    // Compute Merkle root hash
+                    uint256 result = branch.GetHash();
```

Comment:
should error be returned here for invalid MerkleTree structure?

Reply:
@maaku Is there a reason why you presumed `!invalid` on this one? If not I think we should check validity and return error on `invalid=true`.